src/USER-RANN/pair_rann.cpp

```diff
@@ -0,0 +1,1261 @@
+    }
+    elementsp[i] = new char[n];
+    strcpy(elementsp[i],elementwords[i].c_str());
+    net[i].layers = 0;
+    net[i].dimensions = new int[1];
+    net[i].dimensions[0]=0;
+  }
+}
+
+void PairRANN::settings(int narg, char **arg)
+{
+  //read pair_style command in input file
+  if (narg > 0) error->one(FLERR,"Illegal pair_style command");
+}
+
+void PairRANN::coeff(int narg, char **arg)
+{
+  int i,j;
+  map = new int [atom->ntypes+1];
+  if (narg != 3 + atom->ntypes) error->one(FLERR,"Incorrect args for pair coefficients");
+  if (strcmp(arg[0],"*") != 0 || strcmp(arg[1],"*") != 0) error->one(FLERR,"Incorrect args for pair coefficients");
+  nelements = -1;
+  read_file(arg[2]);
+  // read args that map atom types to elements in potential file
+  // map[i] = which element the Ith atom type is, -1 if NULL
+  for (i = 3; i < narg; i++) {
+    if (strcmp(arg[i],"NULL") == 0) {
+      map[i-2] = -1;
+      continue;
+    }
+    for (j = 0; j < nelements; j++) {
+      if (strcmp(arg[i],elements[j]) == 0) break;
+    }
+    if (j < nelements) map[i-2] = j;
+    else error->one(FLERR,"No matching element in NN potential file");
+  }
+  // clear setflag since coeff() called once with I,J = * *
+  int n = atom->ntypes;
+  for (i = 1; i <= n; i++) {
+    for (j = i; j <= n; j++) {
+      setflag[i][j] = 0;
+    }
+  }
+  // set setflag i,j for type pairs where both are mapped to elements
+  // set mass of atom type if i = j
+  int count = 0;
+  for (i = 1; i <= n; i++) {
+    for (j = i; j <= n; j++) {
+      if (map[i] >= 0 && map[j] >= 0) {
+        setflag[i][j] = 1;
+        if (i == j) atom->set_mass(FLERR,i,mass[map[i]]);
+        count++;
+      }
+    }
+  }
+  if (count == 0) error->one(FLERR,"Incorrect args for pair coefficients");
+  for (i=0;i<nelementsp;i++) {
+    for (j=0;j<fingerprintperelement[i];j++) {
+      fingerprints[i][j]->allocate();
+    }
+  }
+  allocated=1;
+}
+
+void PairRANN::read_file(char *filename)
+{
+  FILE *fp;
+  int eof = 0,i,j,k,l;
+  int n,nwords;
+  std::string line,line1;
+  int longline = 4096;
+  char linetemp[longline];
+  char *ptr;
+  bool comment;
+  char str[128];
+  std::vector<std::string> linev,line1v;
+  fp = utils::open_potential(filename,lmp,nullptr);
+  if (fp == nullptr) {error->one(FLERR,"Cannot open RANN potential file");}
+  ptr=fgets(linetemp,longline,fp);
+  strip_comments(linetemp,fp);
+  line=linetemp;
+  while (eof == 0) {
+    ptr=fgets(linetemp,longline,fp);
+    if (ptr == NULL) {
+      if (check_potential()) {
+        error->one(FLERR,"Invalid syntax in potential file, values are inconsistent or missing");
+      }
+      else {
+        eof = 1;
+        break;
+      }
+    }
+    strip_comments(linetemp,fp);
+    line1=linetemp;
+    Tokenizer values = Tokenizer(line,": ,\t_\n");
+    Tokenizer values1 = Tokenizer(line1,": ,\t_\n");
+    linev = values.as_vector();
+    line1v = values1.as_vector();
+    if (linev[0].compare("atomtypes")==0)read_atom_types(linev,line1v);
```

Comment:
why use std::string::compare()? it would be much more readable to use `if (linev[0] == "atomtypes") { `
also the lack of a blank between the conditional and the function calls makes this part of the file difficult to read.